### ISSUE TITLE
Allow CacheManager callbacks to fire synchronously

### ISSUE THREAD

**Issue by aweary** [facebook/relay#1296]
I've implemented a pretty basic `CacheManger` that stores records in a structure that is very similar to `RecordStore`.
#### CacheRecordStore structure

![screen shot 2016-07-22 at 10 59 09 am](https://cloud.githubusercontent.com/assets/6886061/17063116/61bed824-4ffb-11e6-9340-1a2b23914cbf.png)
#### CacheRecord structure

![screen shot 2016-07-22 at 11 00 05 am](https://cloud.githubusercontent.com/assets/6886061/17063146/7bf05722-4ffb-11e6-9ab9-b61aa0525526.png)

Relay has its own internal record store for cached data, `cachedRecords` and `cachedRootCallMap` on `RelayStoreData`. This internal maps are populated by `RelayCachedDataRestorator`.

The issue is that `RelayCachedDataRestorator` is not reading all the records from the `cacheManager`.  It only ever calls `cacheManager.readNode` for what appears to be the root node. I'm using the `relay-starter-kit` to test, so the root node ID is `"VXNlcjox"`. `RelayStoreData._cachedRecords` ends up being populated as:

![screen shot 2016-07-22 at 11 10 52 am](https://cloud.githubusercontent.com/assets/6886061/17063464/ff36a3d8-4ffc-11e6-8a59-b541b93e6526.png)

So the issue here may be that these records are part of a connection. It's interesting that `writeField` is used to write connection records, since it seems that the `dataId` for a connection can change (at least in this example project) so they don't seem to be cacheable.

But the `dataId` for the actual nodes in these edges are consistent, so I would expect Relay to attempt to read those nodes from the `CacheManager`.

**Comment by josephsavona** [comment]
It sounds like your CacheManager implementation may be incomplete. If you're able to share it as a gist or something that would probably be the easiest way to debug.

**Comment by Aweary** [comment]
@josephsavona here's a gist with the current implementation: https://gist.github.com/Aweary/a84664713a6b465c72cd572dc39e698b

Just keep in mind that this a very rough draft. I've implemented a `CacheManager` and `CacheWriter` per the spec defined in `RelayTypes`. 

The fact that `readNode` is only ever invoked for the root node seems like the main problem, maybe I'm returning/storing incomplete data for the record.

**Comment by Aweary** [comment]
@josephsavona here are some potentially useful logs. I'm building Relay from my fork (up to date master) and using that here, so I injected some log into the codepath that seems to be important to restoring cached data. Maybe it will help you more than it's helping me 😄 

![screen shot 2016-07-22 at 1 27 01 pm](https://cloud.githubusercontent.com/assets/6886061/17067184/3668a208-5010-11e6-8e70-16cadfec86f7.png)

`readNode` is the `cacheManager.readNode` method, you'll see it only ever reads the root node.

**Comment by josephsavona** [comment]
@Aweary I'm still looking at your gist, but the first thing that stands out is that the root call map seems to [ignore call values](https://gist.github.com/Aweary/a84664713a6b465c72cd572dc39e698b#file-querymanager-js-L27). Does fixing that help?

**Comment by josephsavona** [comment]
I'm going to close this since this doesn't appear to be an issue with Relay itself. Feel free to continue asking questions here, though, and we'll do our best to answer.

**Comment by Aweary** [comment]
@josephsavona in the simple app I'm testing new with the `callValue` is always an empty string, so I didn't setup a map for them. `readRootCall` just asks that a `dataID` be passed to the callback and it does that successfully, so I don't think that's it. I'll try and get you a repo reproducing this issue, I know it's hard to diagnose with a cursory glance.

**Comment by Aweary** [comment]
@josephsavona here's a repo forked from the relay-starter-kit that demonstrates the issue I'm having: https://github.com/Aweary/relay-cache-manager-issue

I'm pretty sure my `CacheManager` implementation implements the basic interface correctly, according to `RelayTypes` but I could be wrong.

**Comment by josephsavona** [comment]
@Aweary Hmm. The remaining thing that stands out is that record values aren't being serialized with `JSON.stringify`. Other than that, it's going to be a matter of debugging to figure out why the cache read stops early. Your paste of the console logs suggest that the CacheRestorer is trying to follow fields off of the the root node - i'd start debugging there and see why `findRelayQueryLeaves` is unable to continue based on the shape of the data your cache returns.

**Comment by Aweary** [comment]
> The remaining thing that stands out is that record values aren't being serialized with JSON.stringify

@josephsavona the individual records aren't serialized with `JSON.stringify`, but they do get serialized when the `cacheWriter` is serialized, unless I'm misunderstanding you. It does successfully de/serialize the records from what I can see.

The node that's being read from the `cacheManager` **seems** to have the correct shape:

<img width="334" alt="screen shot 2016-07-25 at 11 34 03 am" src="https://cloud.githubusercontent.com/assets/6886061/17109070/097e3dea-525c-11e6-8f32-77f42a773ef2.png">

I'll try to dig into `findRelayQueryLeaves` to see why it seems to return early as far as reading from the cache goes.

**Comment by Aweary** [comment]
@josephsavona quick question, is the `cachedRecords` object passed to `findRelayQueryLeaves` expected to be fully populated when it's passed to `findRelayQueryLeaves`, and subsequently `RelayQueryLeavesFinder`? Or is that done based on the `rangeCalls` array passed in as well (or somewhere else)?

![screen shot 2016-08-01 at 9 24 40 am](https://cloud.githubusercontent.com/assets/6886061/17297103/d8bd48ae-57c9-11e6-9938-d052290c7bf2.png)

I've also noticed `rangeCalls` is `undefined`, so that may be related here.

**Comment by Aweary** [comment]
For what its worth, here's a serialized record of the `store`, `cachedRecords`, `dataID` and `rangeCalls` arguments that are passed to `findRelayQueryLeaves`

<details>
  <summary>Click to expand</summary>
```js

 {
    "store": {
        "_cachedRecords": {
            "VXNlcjox": {
                "__dataID__": "VXNlcjox",
                "__typename": "User",
                "id": "VXNlcjox",
                "widgets": {
                    "__dataID__": "client:17352091011"
                }
            }
        },
        "_cachedRootCallMap": {
            "viewer": {
                "": "VXNlcjox"
            }
        },
        "_queuedRecords": {},
        "_nodeConnectionMap": {},
        "_records": {},
        "_rootCallMap": {},
        "_storage": [
            {},
            {},
            {
                "VXNlcjox": {
                    "__dataID__": "VXNlcjox",
                    "__typename": "User",
                    "id": "VXNlcjox",
                    "widgets": {
                        "__dataID__": "client:17352091011"
                    }
                }
            }
        ]
    },
    "cachedRecords": {
        "VXNlcjox": {
            "__dataID__": "VXNlcjox",
            "__typename": "User",
            "id": "VXNlcjox",
            "widgets": {
                "__dataID__": "client:17352091011"
            }
        }
    },
    "dataID": "VXNlcjox"
}
```
</details>

The record for the root user (`"VXNlcjox"`) has a reference to the widgets connection as `widgets` but I guess I just don't see where/how the cache reader is supposed to know to read the `widgets` field to get the dataID pointing to the connection, which I'm guessing would lead it to read the widget records.

**Comment by marcinbunsch** [comment]
Hi! 

Just wanted to say that I'm also very interested in writing my own Cache Manager which would do something similar. I played around with your repo and I'm also baffled why this would not work, as it seems to conform to the interface. There's some code around the Cache Managers in Relay tests, maybe there will be a suggestion there. I'm not versed enough in the details yet to find it, unfortunately.

I went step by step in the RelayQueryLeavesFinder and it seems to visit RelayQueryField, then runs _visitScalar on the `id` field, which it finds, and there are no more `_pendingNodeStates`. It looks like the fragment fields are not taken into consideration?

**Comment by Aweary** [comment]
Narrowing down the issue a bit more, it seems that `RelayCachedDataRestorator` is not iterating over all the `pendingNodeStates` returned from `findRelayQueryLeaves`. 

`this._state` is marked as `"COMPLETED"` after visiting only the first node state, meaning the `RelayQueryFragment` that is queued up at index `1` is ignored. So now it's matter of identifying where and why state is being marked as completed early.

**Comment by marcinbunsch** [comment]
Ok, so I investigated by going through the whole flow in debugger, step by step, here's what I found. Sorry if it's long, but wanted to recreate all steps I did so you can correct me if I'm wrong anywhere. 

TL;DR

RelayCachedDataRestorator extends RelayCacheProcessor, so when we call `RelayCachedDataRestorator.prototype.traverse`, it calls `RelayCacheProcessor.prototype.visitNode`, which when visitation is complete and there are no more `_pendingRoots` or `_pendingNextStates`, sets the `this._state` to `COMPLETE`, which short circuits `RelayCachedDataRestorator.prototype.traverse` and never allows it to get to the fragment.

Full version:

It all starts with `GraphQLQueryRunner#run`, which receives a querySet, which is a RelayQueryRoot object and has two children:

1) an 'id' field object
2) a RelayFragmentReference object

This runs `runQueries`, where when cacheManager is present, we switch to restoring data from cache (https://github.com/facebook/relay/blob/master/src/legacy/store/GraphQLQueryRunner.js#L244)

At this point we convert the query into requiredQueryMap, with a single key value if `q0: RelayQueryRoot` for the `viewer` field in the query. That RelayQueryRoot object which has two children:

1) RelayQueryField object for 'id'
2) RelayQueryFragment for the view fragment which fetches the widgets. It has the field for `widgets` in children.

With that map, we call `storeData.restoreQueriesFromCache`, which instantiates a changeTracker and profile and calls `restoreQueriesDataFromCache`

This instantiates a `RelayCachedDataRestorator` and calls `restoreQueriesData` on it. This is where visiting and traversal begins.

It starts iterating over the queries and calls `visitRoot` on each of them. In it, it calls `visitIdentifiedRoot`.

In there, it runs `this._store.getDataID`, which returns undefined, and it calls `_RelayCacheProcessor.prototype.visitIdentifiedRoot`. This identifies the root as `viewer*`, stores it in `_pendingRoots` and runs `queueIdentifiedRoot` with that root.

We get to `RelayCacheProcessor.prototype.queueIdentifiedRoot`, which is the first place where we communicate with the cacheManager. (https://github.com/facebook/relay/blob/master/src/store/RelayCacheProcessor.js#L115).

In there we ask the cache do give us the dataID of the root for `viewer`, we return the id of the viewer from the cache.

It builds an array of `pendingRoots`, which holds the one RelayQueryRoot object mentioned before. We start traversing the RelayQueryRoot by calling `RelayCachedDataRestorator#traverse`.

In there `findRelayQueryLeaves` is called. The `RelayQueryLeavesFinder` visits the queryNode. It calls `visitRoot` on it, which calls `traverse`. We traverse the childern of the query node.

First, we run `visitField` on the `id` RelayQueryField. The record is not found, so `_handleMissingData` is called, which pushes the RelayQueryField node to `_pendingNodeStates` and finishes that iteration of the loop.

Secondly, we run `visitFragment` on the RelayQueryFragment. The record is not found, so `_handleMissingData` is called, which pushes the RelayQueryFragment node to `_pendingNodeStates` and finishes that iteration of the loop.

At this point `findRelayQueryLeaves` returns an object with `pendingNodeStates` and within `RelayCachedDataRestorator#traverse` we start iterating over it: https://github.com/facebook/relay/blob/master/src/store/restoreRelayCacheData.js#L242.

We iterate over the pendingNodeStates and call `visitNode` on each. `visitNode` is `RelayCacheProcessor.prototype.visitNode`.

First, we take the `id` RelayQueryField node. This builds an `_pendingNextStates` object with dataID as key and array with one `nextState`. Then we run `RelayCacheProcessor.prototype.queueNode` on the node.

Inside that function, we interact with the cacheManager again. We call `this._cacheManager.readNode`. It returns the object and we mark the node as visited by calling `RelayCachedDataRestorator.prototype.handleNodeVisited` which adds the id to `cachedRecords` and creates in `_changeTracker`.

Then we set a local `pendingNextStates` array and iterate on it. In that loop, we travers the `id` field node using `RelayCachedDataRestorator.prototype.traverse`.

When traversing, we call `findRelayQueryLeaves`, which runs `RelayQueryLeavesFinder.prototype.visitField` on the `id` field. This calls `RelayRecordStore.prototype.getRecordState`. This time, the record is present in `cachedRecords`, so the recordState is `EXISTENT`. Because the record is present, we proceed to visit that field and we call `_visitScalar` on it.

`_visitScalar` finds that the field has data, so it does nothing. `_handleMissingData` is not called.

This results in `findRelayQueryLeaves` to return an empty array of `pendingNodeStates` and the traversal of the `id` RelayQueryField ends

At this point we get back to `RelayCacheProcessor.prototype.queueNode`. There are not `pendingNextStates`, so we finish the loop.

Then, we ask `RelayCacheProcessor` if we're done (https://github.com/facebook/relay/blob/master/src/store/RelayCacheProcessor.js#L172).
This checks if there are any `this._pendingRoots` or any `this._pendingNextStates` and that the state is loading. This resolves to true and we call `_handleSuccess`, which is `RelayCacheProcessor.prototype._handleSuccess`. This sets the `this._state` to `COMPLETED` (https://github.com/facebook/relay/blob/master/src/store/RelayCacheProcessor.js#L228).

In the `_handleSuccess` we call `RelayStoreData.prototype._handleChangedAndNewDataIDs` and set the `readyState` to:

```
readyState.update({
  ready: true,
  stale: true
}, [{ type: 'CACHE_RESTORED_REQUIRED' }]);
```

After all that, we get back to the loop on `pendingNodeStates` in `RelayCachedDataRestorator.prototype.traverse`. However, at this point, `this._state` is `COMPLETED` and this check causes the Restorator to short-circuit: https://github.com/facebook/relay/blob/master/src/store/restoreRelayCacheData.js#L243. That is why we never get to the RelayQueryFragment object.

**Comment by Aweary** [comment]
@marcinbunsch looks like we came essentially to the same conclusion 😄

**Comment by marcinbunsch** [comment]
Yeah, just noticed that after publishing this essay :D

**Comment by marcinbunsch** [comment]
On another note, that looks like a bug in Relay?

**Comment by marcinbunsch** [comment]
Not sure how to fix it, maybe @josephsavona has an idea? Maybe `RelayCachedDataRestorator` should spin up an instance of `RelayCacheProcessor` for each traversal of each `pendingNodeState`?

**Comment by josephsavona** [comment]
Hey there - thanks for all the details. ccing @yuzhi who has worked a lot on cache, one of us will try to take a look.

**Comment by Aweary** [comment]
Thanks @josephsavona! Really appreciate it 👍

**Comment by marcinbunsch** [comment]
I think this can be fixed by adding a `isDone` function to `RelayCachedDataRestorator.prototype` which also checks if `pendingNodeStates` are empty.

**Comment by marcinbunsch** [comment]
Awesome! Thanks, @josephsavona!

**Comment by josephsavona** [comment]
@Aweary or @marcinbunsch - It sounds like this may be a bug. Any chance you could submit a PR with a failing test case?

**Comment by yuzhi** [comment]
Thanks for the super thorough reports!!! I have a guess on what's wrong.  All of our disk caches are actually reading from disk, which is async. Seems like `RelayCachedDataRestorator` expect all the callbacks to be async. Can you guys try adding a setImmediate to see if that solves the problem?

**Comment by marcinbunsch** [comment]
I wrapped the callbacks in `setTimeout(..., 0)` and it worked, thanks! So it's a question if it's a bug or it should work like that by design. Can it be marked somehow in the interface that the callbacks should be async?

Before you wrote that, I tried the following fix in my local copy of `react-relay` which allowed for the loop to finish and the callbacks to be defined in a sync fashion:

```js
RelayCachedDataRestorator.prototype._isDone = function _isDone() {
  var originalIsDone = _RelayCacheProcessor.prototype._isDone.call(this);
  var localIsDone = this.traversalsLeft == 0;
  return originalIsDone && localIsDone;
}

RelayCachedDataRestorator.prototype.traverse = function traverse(node, nextState) {
  this.traversalsLeft = this.traversalsLeft || 0;
  this.traversalsLeft++;
  # ... original traverse code
  this.traversalsLeft--;
};
```

**Comment by marcinbunsch** [comment]
After applying the fix the data is read from cache and rendered, then a request is made to the backend, which re-fetches all of the data and re-renders. Any suggestions what is causing the trip for all data and how we could control the necessity for such trip?

**Comment by yuzhi** [comment]
@marcinbunsch Thanks for testing it out! Sounds like that was the problem. I think it should work for both async and sync reads. This is a case we haven't tested yet with our current implementation. Are either you or @Aweary interested in submitting a pull request to fix it in the `RelayCachedDataRestorator`?

> After applying the fix the data is read from cache and rendered, then a request is made to the backend, which re-fetches all of the data and re-renders.  Any suggestions what is causing the trip for all data and how we could control the necessity for such trip?

Great observation! The current implementation is pretty basic (that's why we haven't really documented it more broadly). It was mostly designed to quickly support product needs for React Native apps at Facebook. We wanted to support offline mode, which works better with a longer TTL, but didn't also want the data we render to be stale when network is available. Because we haven't designed a good API for customizing different level of TTL and most of the apps we were working with wanted to be able to show freshest and most relevant data ASAP, we always send a network request with a disk cache read. Unfortunately, the current data fetching logic does not allow more customizing. Being able to control when to do a network request and when to do a disk read definitely something we want to support better in the future and will actively work on.

**Comment by Aweary** [comment]
I haven't done much work towards actually fixing this, so @marcinbunsch feel free to open a PR to address this! @yuzhi can we re-open this issue, since it turns out that it was a Relay issue?

Thanks for all the help with this everyone 😄

**Comment by Aweary** [comment]
For reference, we've released this basic implementation on npm:
https://github.com/ConciergeAuctions/relay-cache-manager

It's obviously a very naive implementation at the moment, but we'd love to work with anyone who wants to implement an awesome cache manager for Relay.

**Comment by marcinbunsch** [comment]
I took a stab at writing a spec, ended up with this: https://github.com/marcinbunsch/relay/commit/61f80d8a25890f51511082fe18f9a2002687f9d3#diff-0f670f13f2fd3a06b6683e556679ef4aR15

I'm trying to mock `RelayCacheProcessor#visitNode` to see that it gets called for each node in the query. However my jest-fu is not strong and I'm having issues mocking it without the awful `global.calls`. A local variable does not work, because the mocked module gets opened multiple times and these subsequent opens does not share variables. Do you have any suggestions on how to do it properly?

**Comment by josephsavona** [comment]
Resolved right?
